### PR TITLE
fix(grpo): correct DAPO/CISPO/VESPO loss normalization when steps_per_generation != gradient_accumulation_steps

### LINKED ISSUE #5619: DAPO loss is not normalized by gradient_accumulation_steps
Hi, I noticed that the DAPO loss is not normalized by current_gradient_accumulation_steps and am not sure why.  

```

        if self.loss_type in ["grpo", "sapo"]:
            loss = ((per_token_loss * mask).sum(-1) / mask.sum(-1).clamp(min=1.0)).mean()
            normalizer = self.current_gradient_accumulation_steps if mode == "train" else 1.0  # no accum in eval
            loss = loss / normalizer
        elif self.loss_type in ["cispo", "dapo", "vespo"]:
            normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
            loss = (per_token_loss * mask).sum() / normalizer

```
in https://github.com/huggingface/trl/blob/a08e7139f933b770177fc2abc0b43118e26b260b/trl/trainer/grpo_trainer.py#L2547-L2562

### PR DESCRIPTION
**Pull request by 0xadvait**
# What does this PR do?

When `steps_per_generation != gradient_accumulation_steps`, the `dapo`, `cispo` and `vespo` loss types are silently mis-scaled by `gradient_accumulation_steps / steps_per_generation`.

`inputs["num_items_in_batch"]` is computed in `_generate_and_score_completions` as the total completion tokens of the full **generation batch**, which spans `steps_per_generation` micro-steps. The optimizer, however, only accumulates `gradient_accumulation_steps` micro-steps per update. Each micro-step divides its token-loss sum by the generation-batch total, so the loss summed over one accumulation window is normalized by the wrong token count whenever the two differ:

- `steps_per_generation=4`, `gradient_accumulation_steps=2` (the configuration illustrated in the sampler diagram in `grpo_trainer.py`): accumulated loss is **0.5x** the intended token-mean, i.e. gradients are halved
- `steps_per_generation=8`, `gradient_accumulation_steps=32` (the Lite PPO reproduction in [paper_index](https://huggingface.co/docs/trl/main/paper_index), which uses `loss_type="dapo"`): accumulated loss is **4x**, i.e. gradients are quadrupled

Measured on main with a spy on `_compute_loss` (tiny-Qwen2, `loss_type="dapo"`, `num_iterations=1`, `beta=0.0`, no vLLM, so `coef_1 == 1` exactly and the expected loss can be reconstructed from the inputs alone):

```
steps_per_generation=4, gradient_accumulation_steps=2:
  window 0: accumulated_loss=-0.074643 | token-mean over window=-0.149285 | ratio=0.500
  window 1: accumulated_loss=+0.074643 | token-mean over window=+0.149286 | ratio=0.500
steps_per_generation=2, gradient_accumulation_steps=2:
  ratio=1.000 (default-equivalent config, unaffected)
steps_per_generation=2, gradient_accumulation_steps=4:
  ratio=2.0
```

This answers the open question in #5619: `num_items_in_batch` accounts for gradient accumulation only when `steps_per_generation == gradient_accumulation_steps` (the default when neither `steps_per_generation` nor `generation_batch_size` is set). The other loss types are unaffected: `grpo`, `sapo`, `bnpo`, `dr_grpo` and `luspo` normalize within the micro-batch and divide by `current_gradient_accumulation_steps`.

## The fix

Rescale the normalizer to the token count expected in one accumulation window:

```python
normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
if mode == "train":
    normalizer = normalizer * self.current_gradient_accumulation_steps / self.args.steps_per_generation
```

- When `steps_per_generation == gradient_accumulation_steps`, the factor is 1 and behavior is unchanged.
- Otherwise the normalizer becomes the expected token count of one window. Since completions are shuffled across the generation batch before splitting (`shuffle_sequence_dict`), windows are random slices and the expectation is the right scale. Exact per-window counts are unattainable in general because a window can span generation batches that haven't been generated yet (`steps_per_generation < gradient_accumulation_steps`).
- For the `steps_per_generation % gradient_accumulation_steps == 0` case, an exact alternative would be to compute per-window token counts after the split in `_prepare_inputs` and attach them to each slice. I kept the uniform rescale for a minimal, always-applicable change; happy to switch to the exact variant if you prefer it.
- Using `current_gradient_accumulation_steps` (rather than `args.gradient_accumulation_steps`) also handles the final partial accumulation window, consistently with the other loss branches.
- Eval is unchanged: the eval batch is neither split across steps nor accumulated, and `num_items_in_batch` already refers to it.

Note on Liger: #5890 aligns the Liger path with this same `num_items_in_batch` normalizer for `dapo`/`cispo`/`vespo`. If both land, the window rescale should apply there as well; happy to follow up there once #5890 is in.

## Test

`test_dapo_loss_normalized_per_accumulation_window` reconstructs the expected loss from captured inputs (`per_token_loss == -advantages` at the first iteration with `beta=0.0` and no vLLM) and is parametrized over `(steps_per_generation, gradient_accumulation_steps)` in `(4,2)`, `(2,2)`, `(2,4)`. The `(4,2)` and `(2,4)` cases fail without this fix; `(2,2)` passes before and after, guarding the default behavior.

Fixes #5619

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. -> #5619
- [x] Did you make sure to update the documentation with your changes? (no user-facing docs describe this normalizer; an explanatory comment was added at the fix site)
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec (authored the DAPO loss type in #3938 and replied in #5619) -- also relevant to @kashif's #5890 (Liger normalizer alignment).

---

> [!NOTE]
> **Medium Risk**
> Changes RL loss scaling for DAPO/CISPO/VESPO training when generation/accumulation steps differ, directly affecting gradient magnitude; default-equal configs are unchanged and eval is unaffected.
> 
> **Overview**
> Fixes **mis-scaled policy gradients** for `loss_type` **`dapo`**, **`cispo`**, and **`vespo`** when **`steps_per_generation`** and **`gradient_accumulation_steps`** differ.
> 
> Those losses divide by **`num_items_in_batch`**, which counts completion tokens across the full **generation batch** (`steps_per_generation` micro-steps), while the optimizer only sums **`gradient_accumulation_steps`** micro-batches per update. That mismatch skews the accumulated loss (e.g. **0.5×** or **4×** the intended token-mean in common configs).
> 
> In **`_compute_loss`** (main **`GRPOTrainer`** and experimental **`gspo_token`**), training now rescales the normalizer by **`current_gradient_accumulation_steps / steps_per_generation`**; when the two settings match, behavior is unchanged. **Eval** is untouched.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 93929baef2bc635d07460769cef3ce3329ea1c30. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by behroozazarkhalili** [comment]
I traced #5619 independently against current `main` and can confirm both the bug and that this fix is correct.

**The bug is real.** In `_compute_loss`, the `grpo`/`bnpo`/`dr_grpo` branches divide by `current_gradient_accumulation_steps`, but the `cispo`/`dapo`/`vespo` branch normalizes only by `num_items_in_batch / num_processes`. `num_items_in_batch` counts the completion tokens of the whole generation batch (which spans `steps_per_generation` micro-steps), while the optimizer only accumulates `gradient_accumulation_steps` micro-step losses before stepping. So the accumulated loss carries an extra factor of `gradient_accumulation_steps / steps_per_generation`, which is 1 only when the two coincide. That matches @0xadvait's measured 0.5x / 1.0x / 2.0x for `(spg=4, gas=2)`, `(spg=2, gas=2)`, `(spg=2, gas=4)`, and it explains why the earlier "num_items_in_batch already accounts for accumulation" reasoning holds for the default `spg == gas` case but not in general.

**The fix is correct on the points I checked:**
- Multiplying the normalizer by `current_gradient_accumulation_steps / steps_per_generation` exactly cancels the `gas / spg` mis-scale, so the accumulated loss becomes the token-mean over one accumulation window.
- Using `current_gradient_accumulation_steps` (rather than `args.gradient_accumulation_steps`) is the right choice: transformers sets it to `len(batch_samples)`, so it is correct for a partial final accumulation window too, and it matches what the other loss branches already use.
- The `if mode == "train"` guard is right: in eval the batch is not split across `steps_per_generation` and not accumulated, so the normalizer should stay as-is.
- It covers `cispo` and `vespo` as well as `dapo`, since all three share the branch.

The regression test reconstructs the expected loss from the inputs under `num_iterations=1, beta=0`, which is a clean way to pin the scale. Looks good to me.

cc @qgallouedec

**Comment by kashif** [comment]
FYI #5890 is now ready for review and bumps the liger-kernel pin to 0.8.1, so the Liger normalizer-rescale you mentioned should be doable now. Happy to take it on here (or as a follow-up) — let me know which you'd prefer.

**Comment by behroozazarkhalili** [comment]
Thanks @kashif. Small correction first: the Liger rescale idea was not mine, I only reviewed the non-Liger fix here, but I did check the math so happy to weigh in on how the two interact, because there is a coordination point worth being explicit about.

Verified against current `main` plus both diffs. Writing the effective normalizer as the divisor in `loss = (per_token_loss * mask).sum() / normalizer` (train mode):

```python
# main (today)
non_liger = num_items_in_batch / num_processes                      # no grad-accum factor
liger     = current_gradient_accumulation_steps                     # and num_items_in_batch is not passed to the kernel

# after #6024 (non-Liger dapo/cispo/vespo)
non_liger = num_items_in_batch / num_processes * current_gradient_accumulation_steps / steps_per_generation

# after #5890 (Liger dapo/cispo/vespo: kernel normalizes by num_items_in_batch, trainer drops the /grad_accum divide)
liger     = num_items_in_batch / num_processes                      # no gas/spg factor
```

So if both land as they stand, the Liger and non-Liger losses diverge again by exactly `current_gradient_accumulation_steps / steps_per_generation` whenever `steps_per_generation != gradient_accumulation_steps`. #5890 matched Liger to the pre-#6024 non-Liger path, and #6024 then moves that path, so the same class of mismatch reappears on the Liger side, this time driven by `steps_per_generation`.

The reason is the same for both paths: `num_items_in_batch` counts the whole generation batch (`steps_per_generation` micro-steps) while the optimizer only accumulates `gradient_accumulation_steps` of them, so the Liger path needs the identical `current_gradient_accumulation_steps / steps_per_generation` rescale to stay in parity.

On here-vs-follow-up: I would keep #6024 scoped to the non-Liger path (it is verified correct on its own) and add the same rescale in #5890, so both paths end at:

```python
num_items_in_batch / num_processes * current_gradient_accumulation_steps / steps_per_generation
```

Whichever merges second should carry the other's factor. One heads-up: #5890's `test_compute_liger_loss_dapo_normalizer` currently asserts the final loss equals the raw Liger output (no rescale), so that assertion will need updating once the `current_gradient_accumulation_steps / steps_per_generation` factor is added.

Happy to re-run the spy-on-`_compute_loss` check across both paths once they are on a branch together to confirm they match numerically. cc @qgallouedec

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -3137,7 +3137,15 @@ def _compute_loss(self, model, inputs):
...
         elif self.loss_type in ["cispo", "dapo", "vespo"]:
+            # `num_items_in_batch` counts the completion tokens of the full generation batch, which spans
+            # `steps_per_generation` micro-steps, while gradients are only accumulated over
+            # `gradient_accumulation_steps` micro-steps. When the two differ, normalizing every micro-step by the
+            # generation-batch token count scales the accumulated loss by
+            # gradient_accumulation_steps / steps_per_generation. Rescale the normalizer to the token count
+            # expected in one accumulation window. See https://github.com/huggingface/trl/issues/5619.
             normalizer = inputs["num_items_in_batch"].clamp(min=1.0) / self.accelerator.num_processes
+            if mode == "train":  # in eval, the batch is neither split across steps nor accumulated
+                normalizer = normalizer * self.current_gradient_accumulation_steps / self.args.steps_per_generation
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:3159]: Fixed in 1358cd2b, and agreed this is the repo's duplication policy at work. The rescale is now applied to the `dapo` branch of the duplicated `_compute_loss` in `trl/experimental/gspo_token/grpo_trainer.py`, matching the main trainer:

```python
normalizer = inputs["num_items_in_batch"].clamp(min=1.0) / self.accelerator.num_processes
if mode == "train":  # in eval, the batch is neither split across steps nor accumulated
    normalizer = normalizer * self.current_gradient_accumulation_steps / self.args.steps_per_generation
```

Scope check: `gspo_token` only supports the `dapo` loss type (it raises for the others), and it is the only experimental copy carrying this `num_items_in_batch` normalizer signature. `gmpo` has no `dapo` branch and no other experimental trainer matches, so this one file was the whole gap. cc @qgallouedec @kashif

**Comment by qgallouedec** [comment]
Thanks both, merging the two normalizer hunks; I reproduced the 0.5×/1×/2× scaling for dapo, cispo and vespo.

@behroozazarkhalili you're right on the Liger side. I'm merging this one first, #5890 already carries the same factor, so the two paths should line up. @kashif nothing needed on your end beyond a rebase.

I dropped the test. It reconstructs the expected loss from `per_token_loss == -advantages`, which only holds when `old_per_token_logps is None`. In the `(4, 2)` case `gradient_accumulation_steps % generate_every != 0`, so `old_per_token_logps` comes from a separate chunked `no_grad` forward and `coef_1` drifts ~1e-3 from 1, above `rtol=1e-4`. It fails here even at `learning_rate=0.0`, and the two passing cases are exactly the ones where `coef_1` is 1 by construction.
The assertion also restates the implementation formula, so it would pass for any code computing that expression. The property worth pinning is that the loss summed over one accumulation window is invariant to the spg/gas split. Not sure if it's worth it.

Thanks a lot guys!

**Review by qgallouedec — APPROVED** [review]
